### PR TITLE
New: Add support for duplicate object literal properties

### PR DESCRIPTION
**Pull request by FredKSchott**
- Adds support for duplicate object properties in ES 6 with the `duplicateObjectLiteralProperties` feature. 
- Continues to forbid duplicate object literals in strict mode otherwise
- Special rules for the `__proto__` property

This feature doesn't exist in esprima-fb yet, so we couldn't use it to help generate tests. Instead all tests were created by this branch and then verified by me. It looks good to me, but a few more pairs of eyes couldn't hurt.

### CODE REVIEW

File: lib/features.js [hunk 1]
```diff
@@ -66,6 +66,9 @@ module.exports = {
     // enable parsing of shorthand object literal properties
     objectLiteralShorthandProperties: false,
 
+    // Allow duplicate object literal properties (except '__proto__')
+    duplicateObjectLiteralProperties: false,
+
```
Comment (nzakas) [inline lib/features.js]: I think it would be better to name `objectLiteralDuplicateProperties` to match the other flags that apply to object literals.

File: espree.js [hunk 2]
```diff
@@ -2459,20 +2460,29 @@ function parseObjectInitialiser() {
...
             if (property.key.type === astNodeTypes.Identifier) {
                 name = property.key.name;
             } else {
                 name = toString(property.key.value);
             }
 
-            /*eslint-disable no-nested-ternary*/
-            kind = (property.kind === "init") ? PropertyKind.Data : (property.kind === "get") ? PropertyKind.Get : PropertyKind.Set;
-            /*eslint-enable no-nested-ternary*/
+            if (property.kind === "init") {
```
Comment (nzakas) [inline espree.js]: Nice!

File: lib/features.js (hunk 1 above)
Comment (FredKSchott) [inline lib/features.js]: makes sense, done.

File: tests/fixtures/ecma-features/duplicateObjectLiteralProperties/proto-property.src.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+"use strict";
+
+var proto = {};
+
+var x = {
+	__proto__: proto,
```
Comment (nzakas) [inline tests/fixtures/ecma-features/duplicateObjectLiteralProperties/proto-property.src.js]: Can you add another test using the string "__proto__" twice? It should not throw an error in that case.

**Comment by nzakas** [comment]
This looks really good. I'd like to see tests using strings as identifiers. For instance, you can have duplicate "__proto__" properties if they are strings. So I think if you duplicate your current tests and use string property names, that should give us good enough coverage.

Also, be sure to update the readme with the new option.

**Comment by FredKSchott** [comment]
Agreed, tests should be added for different kinds of property names. But I'm not sure if the ES 6 spec matches the behavior you're describing. See: http://people.mozilla.org/~jorendorff/es6-draft.html#sec-object.prototype.__proto__

For ***"\_\_proto\_\_ Property Names in Object Initializers"*** it says:
> It is a Syntax Error if PropertyNameList of PropertyDefinitionList contains any duplicate entries for "\_\_proto\_\_" and at least two of those entries were obtained from productions of the form `PropertyName : AssignmentExpression` .

Where `PropertyName` can be a `LiteralPropertyName` defined as:
```
PropertyName[Yield,GeneratorParameter] :
  LiteralPropertyName
  [+GeneratorParameter] ComputedPropertyName
  [~GeneratorParameter] ComputedPropertyName[?Yield]

LiteralPropertyName :
  IdentifierName
  StringLiteral
  NumericLiteral
```

So judging by the spec, `__proto__` duplicates will result in a syntax error if duplicates are defined as IdentifierName, StringLiteral, or ComputedPropertyName.

**Comment by nzakas** [comment]
Looking at this closer, we are both wrong. :) I had sent a question about this to the mailing list and misunderstood a piece:
https://esdiscuss.org/topic/question-regarding-duplicate-proto-properties

`["__proto__"] : foo` is the thing that is treated differently. You cannot set a prototype using a computed property. This is reflected in step 6 of https://people.mozilla.org/~jorendorff/es6-draft.html#sec-__proto__-property-names-in-object-initializers.

So really the special case is with computed properties, not strings.

**Comment by FredKSchott** [comment]
Ah you're right, thats exactly what the numbered steps are explaining. That's surprisingly satisfying...

The spec definitions themselves (not including those number steps) still seem to suggest that all PropertyNames (including computed) are valid for the `__proto`` special case. Do you think that is worth bringing up on es-discuss?

**Comment by nzakas** [comment]
I don't think so. B.3.1 is the place in the spec that fully defines this behavior, including the various exceptions. It's referred to in other sections as well. The most relevant is the runtime semantics of PropertyDefinitionEvaluation: https://people.mozilla.org/~jorendorff/es6-draft.html#sec-object-initializer-runtime-semantics-propertydefinitionevaluation

**Comment by FredKSchott** [comment]
Gotcha, I'm still unfamiliar with this spec / reading language specifications in general :)

**Comment by nzakas** [comment]
No problem. So I think tasks here are too as a couple of tests with string property names and add a test that has both computed and duplicate properties enabled to try `["__proto__"]` (see ecma-features-mix directory for examples.

**Comment by FredKSchott** [comment]
More tests added, let me know if there are any important cases that I missed

**Comment by nzakas** [comment]
Lgtm. Thanks!

**Comment by FredKSchott** [comment]
Follow up: I was still confused by the wording of the spec, so I ended up composing my confusion a bit more clearly and posting it to es-discuss. Looks like we were right on the behavior and the confusion was caused by an error in the spec.

the thread in question, for the curious: https://esdiscuss.org/topic/question-regarding-duplicate-computed-proto-properties

**Comment by nzakas** [comment]
Nice!
